cpp/src/arrow/array/data.h

```diff
@@ -33,6 +33,19 @@
 namespace arrow {
 
 class Array;
+struct ArrayData;
+
+namespace internal {
+// ----------------------------------------------------------------------
+// Null handling for types without a validity bitmap
+
+ARROW_EXPORT bool IsNullSparseUnion(const ArrayData& data, int64_t i);
+ARROW_EXPORT bool IsNullDenseUnion(const ArrayData& data, int64_t i);
+ARROW_EXPORT bool IsNullRunEndEncoded(const ArrayData& data, int64_t i);
+
+ARROW_EXPORT bool UnionMayHaveLogicalNulls(const ArrayData& data);
+ARROW_EXPORT bool RunEndEncodedMayHaveLogicalNulls(const ArrayData& data);
```

Comment:
What prevents these functions from being `.cc` only function in an anonymous namespace?  It seems like the only need is because we have functions later in `data.h` that could be in a `.cc` file.

Reply:
They are called from `IsNull` and `IsValid` which are inline and having the switch on the type ID inlined allows the compiler to inline the highly predictable branches into the loos bodies from which these functions are called.

Reply:
I could introduce an `IsValidFallback` function that handles the case when the validity buffer is `NULLPTR`, then all these could become `.cc` only.

Reply:
I think it's ok how it is.  I just don't ever have a good sense for when it's justified for leaving something in the header file for performance reasons.  Typically I think it would be nice for there to be some benchmark to point to.  Otherwise "it seems like this is important for performance" becomes too subjective of a criteria to apply.

Reply:
The existing benchmarks indicated some regression (maybe due to union arrays being part of `IsNull` benchmarks) so I will investigate more carefully.